streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1256,10 +1261,25 @@ private <VO, VR> KStream<K, VR> doStreamTableJoin(final KTable<K, VO> table,
         final NamedInternal renamed = new NamedInternal(joinedInternal.name());
 
         final String name = renamed.orElseGenerateWithPrefix(builder, leftJoin ? LEFTJOIN_NAME : JOIN_NAME);
+
+        Optional<TimeOrderedKeyValueBuffer<K, V, V>> buffer = Optional.empty();
+
+        if (joined.gracePeriod() != null) {
```

Comment:
Shouldn't this be
```suggestion
        if (joined.gracePeriod() != null && joined.gracePeriod() > 0) {
```
In the KIP discussion we said that a grace period of zero is equal to no grace period.

Note, some conditions later on need to be changed if you agree with this change, like for example `gracePeriod.isPresent() ^ buffer.isPresent()`.

Reply:
I see what you mean but we actually were going to create the store and leave it empty for zero duration. The point was to make it easier to change the grace period if desired so the store isn't getting created and destroyed. Something about making it more backward compatible.

Reply:
So we are going to make a store even when the grace period is zero. This should make changing the grace period to and from zero smoother. So if someone set the grace period to zero then we can evict everything from the buffer even after the grace period is changed.

Reply:
> So if someone set the grace period to zero then we can evict everything from the buffer even after the grace period is changed.

Are you saying that if I set the grace period > 0 and I run my application, then I stop my application and set the grace period == 0, and restart my application, my application will process records from my first run?
Why should removing the grace period (grace period == null) behave differently?
Actually we do not support topology updates. However, we can at least make equivalent changes behave equivalently.

Reply:
Another thought here: 
What happens if I specify a grace period > 0 and run my application for a while. Then, I remove the grace period and again run my application for a while. Finally, I specify a grace period > 0 or even == 0 and again run my application for a while. Does my application in the last run process records that the first run wrote into the buffer? (Assuming the task is always assigned to the same node).

Reply:
What I am trying to say is that we should try to put simplicity over partial support for topology updates.

Reply:
This seems to be the main sticking point of the PR. 

The thought was that this that we would need some way to stop using the grace period without loosing records, but you bring up a good point.